smithy-utils/src/main/java/software/amazon/smithy/utils/CodeWriter.java

```diff
@@ -73,13 +74,25 @@
  * formatters:
  *
  * <ul>
- *     <li>{@code L}: Outputs a literal value of an {@code Object} using
+ *     <li>{@code L} (literal): Outputs a literal value of an {@code Object} using
  *     the following implementation: (1) A null value is formatted as "".
  *     (2) An empty {@code Optional} value is formatted as "". (3) A non-empty
  *     {@code Optional} value is recursively formatted using the value inside
  *     of the {@code Optional}. (3) All other valeus are formatted using the
  *     result of calling {@link String#valueOf}.</li>
- *     <li>{@code S}: Adds double quotes around the result of formatting a
+ *
+ *     <li>{@code C} (call): Runs a {@link Runnable} argument that is expected to
+ *     write to the same writer. Any text written to the CodeWriter inside of the
+ *     Runnable is used as the value of the argument. Note that a single trailing
+ *     newline is removed from the captured text.
+ *
+ *     <pre>{@code
+ *     CodeWriter writer = new CodeWriter();
+ *     writer.write("Hello, $C.", () -> writer.write("there"));
```

Comment:
interesting—I wonder if this is a potential for bugs + limiting. It seems like it might be possible to end up accidentally calling a different writer since you're relying on capturing the writer in a closure. 

It also makes it a bit tricker because you _can't_ create a Runnable without also having access to the base code writer at the same time.

The alternative being `(CodeWriter) -> Unit` where `CodeWriter` handles passing itself through

Reply:
Yeah I thought about that too. We could support both `Runnable` and `Consumer<CodeWriter>`. The issue there is that I don't think it can be `Consumer<T extends CodeWriter>`since we only know it's a CodeWriter and not some special subtype.

Reply:
ah I see, yeah you'd need to add a self-type to code writer, but I don't think that could be done in a backwards compatible way.

Reply:
actually I think you could do it with a refactor like this [note: I don't really know Java]
```java
abstract class BaseCodeWriter<Self extends BaseCodeWriter<Self>> {
  // all of the codewriter functionality actually goes here, but now you can name `Self`
}

class CodeWriter extends BaseCodeWriter<CodeWriter> {
}

class RustWriter extends BaseCodeWriter<RustWriter> { 
  // rust specific functionality
}
```

Reply:
That would work, but turns out isn't necessary for the problem I was describing. We can just do a cast in Java and call the Consumer anyways. Your solution would be useful for cleaning up a different problem of ugly parameters, but would need more helper methods on CodeWriter to essentially be an identity function that makes a safe cast of a Runnable or Consumer.

For example, this:
```
CodeWriterSubtype w = new CodeWriterSubtype();
w.write("$C", (Consumer<CodeWriterSubtype>) writer -> writer.write2("Hello!"));
```

Could become:
```
CodeWriterSubtype w = new CodeWriterSubtype();
w.write("$C", w.call(writer -> writer.write2("Hello!")));
```

Slightly less verbose, but it requires the use of a helper function, and requires a big refactor of CodeWriter. I'm thinking it would be prudent to leave this out for now.